ssz/sedes/serializable.py

```diff
@@ -0,0 +1,425 @@
+    def __init__(self, *args, **kwargs):
+        validate_args_and_kwargs(args, kwargs, self._meta.field_names)
+        field_values = merge_kwargs_to_args(args, kwargs, self._meta.field_names)
+
+        # Ensure that all the fields have been given values in initialization
+        if len(field_values) != len(self._meta.field_names):
+            raise TypeError(
+                'Argument count mismatch. expected {0} - got {1} - missing {2}'.format(
+                    len(self._meta.field_names),
+                    len(field_values),
+                    ','.join(self._meta.field_names[len(field_values):]),
+                )
+            )
+
+        for value, attr in zip(field_values, self._meta.field_attrs):
+            setattr(self, attr, make_immutable(value))
+
+    def as_dict(self):
+        return dict(
+            (field, value)
+            for field, value
+            in zip(self._meta.field_names, self)
+        )
+
+    def __iter__(self):
+        for attr in self._meta.field_attrs:
+            yield getattr(self, attr)
+
+    def __getitem__(self, index):
+        if isinstance(index, int):
+            attr = self._meta.field_attrs[index]
+            return getattr(self, attr)
+        elif isinstance(index, slice):
+            field_slice = self._meta.field_attrs[index]
+            return tuple(getattr(self, field) for field in field_slice)
+        elif isinstance(index, str):
+            return getattr(self, index)
+        else:
+            raise IndexError("Unsupported type for __getitem__: {0}".format(type(index)))
+
+    def __len__(self):
+        return len(self._meta.fields)
+
+    def __eq__(self, other):
+        return isinstance(other, Serializable) and hash(self) == hash(other)
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        # The hash() builtin is not stable across processes
+        # (https://docs.python.org/3/reference/datamodel.html#object.__hash__), so we do this here
+        # to ensure pickled instances don't carry the cached hash() as that may cause issues like
+        # https://github.com/ethereum/py-evm/issues/1318
+        state['_hash_cache'] = None
+        return state
+
+    _hash_cache = None
+
+    def __hash__(self):
+        if self._hash_cache is None:
+            self._hash_cache = hash(tuple(self))
+
+        return self._hash_cache
+
+    @classmethod
+    def serialize(cls, obj):
+        # Make sure that the obj is instance of class before serializing
+        if not isinstance(obj, cls):
+            raise SerializationError(
+                f"Can only serialize objects which are instances of {cls.__name__} class",
+                obj
+            )
+
+        serialized_string = b"".join(
+            field_sedes.serialize(getattr(obj, field_name))
+            for field_name, field_sedes in cls._meta.fields
+        )
+
+        if len(serialized_string) >= 2 ** (CONTAINER_PREFIX_LENGTH * 8):
```

Comment:
We should probably set `2 ** (CONTAINER_PREFIX_LENGTH * 8)` to a module level constant so we don't incur the overhead of computing it everytime.